Add Discord Rich Presence support - Launcher

L-e-x-o-n opened the pull request:
This is the launcher side PR for adding Discord Rich Presence support for BAR.
Statuses won't be shown until both Chobby and launcher changes are merged and active, but they can be merged independently without new issues.
[Chobby PR](https://github.com/beyond-all-reason/BYAR-Chobby/pull/578)

# What is DIscord Rich Presence?
Discord Rich Presence makes it possible to display statuses and details of games and programs that support it.
If you use DIscord you probably saw people that have statuses below their names: "Playing *game*", "Listening to _Spotify song_", "Streaming" etc. Rich Presence makes this possible and currently, unless they added BAR in their registered games section in Discord settings, people playing BAR wouldn't have the "Playing Beyond All Reason" status.

Activity status will only be shown for people who enable it in Discord settings under Activity Settings>Activity Privacy>Activity Status section.

Rich Presence supports more than just showing a basic one liner. For now I added more details, but in the future it is possible to expand it with the ability to send game invites and join games directly from Discord.

# Example
![image](https://github.com/beyond-all-reason/spring-launcher/assets/44340857/079539ef-6e96-473b-88a2-137e3556670c)
![image](https://github.com/beyond-all-reason/spring-launcher/assets/44340857/1d71854d-c2b4-4f78-9c3a-255e286a4ca0)

# Details shown
- Game title
- Map name
- Map image, thanks to @p2004a
- Status: playing, spectating, in menu, in (running) lobby, watching replay, skirmish
- Time elapsed
- Play button that leads to https://www.beyondallreason.info/

# What this PR does
- Adds [Discord.js RPC Extension](https://github.com/discordjs/RPC)
- New extension that receives `DiscordSetActivity` message from Chobby and sets Discord activity details
- Adds `config.json` to [.gitignore](https://github.com/beyond-all-reason/spring-launcher/blob/master/.gitignore)

L-e-x-o-n committed:
Add Discord Rich Presence support

p2004a commented on src/exts/discord_integration.js:
```diff
@@ -0,0 +1,71 @@
+const DiscordRPC = require("discord-rpc");
+
+// Application ID from https://discord.com/developers/applications/<clientId>/
+const clientId = '1185990483143569448';
```
where does it come from? If you had to register it, then this should be a config.json, and all bar specific options should be configured like that.

I would like to keep this repo without any BAR specific things, so it could be used by other Recoil games.

L-e-x-o-n replied:
This is a generated application/client ID you get when creating an application on https://discord.com/developers/applications. 

Should I add `https://maps-metadata.beyondallreason.dev/latest/discordPresenceThumb/redir.${encodeURIComponent(command.details)}.1024.jpg there as well?

p2004a replied:
Yeah, some sort of string that can be a key or just url to template.

L-e-x-o-n replied:
Done

p2004a commented on src/exts/discord_integration.js:
```diff
@@ -0,0 +1,71 @@
+DiscordRPC.register(clientId);
+const rpc = new DiscordRPC.Client({ transport: 'ipc' });
+
+bridge.on('DiscordSetActivity', async command => {
```
could you please add some comment above those function about what is the expected format of command? What properties does it take etc.

L-e-x-o-n replied:
Done

L-e-x-o-n committed:
Added config

p2004a commented on src/exts/discord_integration.js:
```diff
@@ -0,0 +1,94 @@
+}
+
+// Application ID from https://discord.com/developers/applications/<applicationId>/
+const applicationId = "1185990483143569448";
```
I think you missed to delete this

L-e-x-o-n replied:
Missed that one, thanks.

p2004a commented on src/exts/discord_integration.js:
```diff
@@ -0,0 +1,94 @@
+DiscordRPC.register(applicationId);
+
+bridge.on('DiscordSetActivity', async command => {
+    // command
```
maybe also add link to https://discord.com/developers/docs/rich-presence/how-to#updating-presence-update-presence-payload-fields there it's actually explained what those mean, and not just what type it is, I see you add comment to each assignment below, but you could just add them here instead and make it more concise.

L-e-x-o-n replied:
Will do

p2004a commented on src/exts/discord_integration.js:
```diff
@@ -0,0 +1,94 @@
+    // Name of the uploaded image for the large profile artwork
+    // Minimap Image URL
+    const largeImageKey = command.details ?
+        String(config.discord_rich_presence.minimap_link).replace("<map>", encodeURIComponent(command.details)) :
```
You don't need to and shouldn't call the `String`, just do .replace on the `minimap_link`.

L-e-x-o-n replied:
I am using String in case `config.discord_rich_presence.minimap_link` is not defined in conig and is null.

p2004a replied:
If it's null, why set it as 'null' string?

p2004a commented on src/exts/discord_integration.js:
```diff
@@ -0,0 +1,94 @@
+        config.discord_rich_presence.large_image_text_default;
+
+    // Name of the uploaded image key for the small profile artwork
+    const smallImageKey = command.smallImageKey ? command.smallImageKey : config.discord_rich_presence.small_image_key_default;
```
You could use the null coalescing operator to make it cleaner https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing

And also, you could assign directly in the object below, no need to assign to those temporary variables.

L-e-x-o-n replied:
Thanks, I didn't know this existed.

p2004a commented on src/exts/discord_integration.js:
```diff
@@ -0,0 +1,94 @@
+    // Both playerCount and maxPlayerCount have to be > 0 to avoid errors
+    let playerCount;
+    let maxPlayerCount;
+
```
nit: would drop this single empty line

p2004a commented on src/exts/discord_integration.js:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    // Party id (battle id for now)
+    const partyId = String(command.partyId)
```
to convert to string, it's `.toString()` method, calling `String` is not idiomatic

L-e-x-o-n replied:
I am using String in case `command.partyId` is null.

p2004a replied:
Same as other comment, if it's null, why set it as 'null' string?

p2004a commented on src/exts/discord_integration.js:
```diff
@@ -0,0 +1,94 @@
+        partyId: partyId,
+        buttons: [{
+            label: "Play",
+            url: "https://www.beyondallreason.info/",
```
BARism

L-e-x-o-n replied:
Replaced

p2004a commented on src/exts/discord_integration.js:
```diff
@@ -0,0 +1,94 @@
+    //     smallImageText - string
+    // }
+
+    if (!rpc) {
```
when can this happen when constructor is called above?

L-e-x-o-n replied:
Never :) I reworked that part

p2004a commented on src/exts/discord_integration.js:
```diff
@@ -0,0 +1,94 @@
+const { config } = require('../launcher_config');
+const DiscordRPC = require("discord-rpc");
+
+if (config) {
```
why this if? this is never null

L-e-x-o-n replied:
Fixed

p2004a reviewed (commented):
Some minor things, overall looks good so I will be able to merge soon.

L-e-x-o-n committed:
Improvements

L-e-x-o-n commented:
I made changes based on review, reworked login part.
I kept `String()` instead of using `.toString()` because it's possible for those strings to be `null`. Is there a better solution?

p2004a commented on src/exts/discord_integration.js:
```diff
@@ -0,0 +1,77 @@
+
+async function TryToLogin() {
+    try {
+        await rpc.login({ clientId: applicationId });
```
Reading discord-rpc code, this will never reconnect on failure, I believe you have to create a new client object. Also it won't reconnect if you stop and start discord. Can you please confirm/test for that? And is that intended?

p2004a commented on src/exts/discord_integration.js:
```diff
@@ -0,0 +1,77 @@
+        maxPlayerCount = command.maxPlayerCount;
+    }
+    
+    rpc.setActivity({
```
You await here, catch exceptions, and log them as warnings.

p2004a commented on src/exts/discord_integration.js:
```diff
@@ -0,0 +1,77 @@
+    // https://discord.com/developers/docs/rich-presence/how-to#updating-presence-update-presence-payload-fields
+    // {
+    //     state : string - Playing, spectating, in menu, in lobby, watching replay...
+    //     details : string - Map name
```
btw why is it details a not "mapName" ?

L-e-x-o-n replied:
I was trying to use the same names as in Discord's documentation.

L-e-x-o-n replied:
I realised I wasn't consistent in following this so I renamed it to mapName and partyId to battleId

p2004a commented on src/exts/discord_integration.js:
```diff
@@ -0,0 +1,77 @@
+    if(command.playerCount > 0 && command.maxPlayerCount > 0) {
+        playerCount = command.playerCount;
+        maxPlayerCount = command.maxPlayerCount;
+    }
```
Instead of doing it like that, you could create object, and then add more properties to it, this will also help with the party null. Like

```
const request = {
   // all the known fields
};
if(command.playerCount > 0 && command.maxPlayerCount > 0) {
    request.partySize = command.playerCount;
    request.partyMax = command.maxPlayerCount;
}
// (...)
await rpc.setActivity(request);
```

L-e-x-o-n committed:
Improvements based on review

L-e-x-o-n committed:
Forgotten ;

L-e-x-o-n committed:
Improvements, reconnec working

L-e-x-o-n committed:
Improvements to reconnect

p2004a committed:
Fix crash when property not defined

Fails with

```
Failed to load extension discord_integration.js with error: TypeError: Cannot read properties of undefined (reading 'application_id')
```

without it

p2004a committed:
Fix eslint linting errors

p2004a committed:
Some small stylistic changes

p2004a commented:
Ok, I've pushed a few small fixes, tested and it looks to be working nicely :)

I will also need to make a few changes to the flatpak distribution to make sure it also works correctly there, according to this guide: https://github.com/flathub/com.discordapp.Discord/wiki/Rich-Precense-(discord-rpc)#flatpak-applications
